test/System.Net.Http.Formatting.Test/HttpMessageContentTests.cs

```diff
@@ -76,6 +76,21 @@ private static async Task ValidateRequest(HttpContent content, bool containsEnti
         private static async Task ValidateResponse(HttpContent content, bool containsEntity)
         {
             Assert.Equal(ParserData.HttpResponseMediaType, content.Headers.ContentType);
+
+#if NET6_0_OR_GREATER
+            // Test initialization leaves the inner HttpContent (EmptyContent or StreamContent) w/ a non-null but
+            // unseekable Stream (EmptyReadStream). HttpMessageContent believes the message has content but is unable
+            // to calculate the message length correctly in this case. Overriding the Content e.g. w/ a StringContent
```

Comment:
> HttpMessageContent believes the message has content but is unable to calculate the message length correctly in this case

What does HttpMessageContent do when using HttpClient for real?  It doesn't produce seekable streams, even on .NET Framework.

If I'm looking at the right code, it just says it can't get the length in that case:
https://github.com/aspnet/AspNetWebStack/blob/169d95fc063feb401736e85aab1aa26518e78a92/src/System.Net.Http.Formatting/HttpMessageContent.cs#L225-L229

Doesn't that just mean this test is buggy, in that it was testing artificial behaviors produced by the test rather than testing what happens in real use?

Reply:
> > HttpMessageContent believes the message has content but is unable to calculate the message length correctly in this case
> 
> What does HttpMessageContent do when using HttpClient for real? Unless you buffer, it doesn't produce seekable streams, even on .NET Framework.

There isn't much e2e testing in this repository and there are too many cases for me to be sure of the answer. It also likely depends on the exact order of operations in user code. I see for example that almost any interaction w/ the read stream in `HttpContent` will buffer it. See https://github.com/microsoft/referencesource/blob/master/System/net/System/Net/Http/HttpContent.cs#L358-L373 and the use of that method.
 
> If I'm looking at the right code, &hellip;
>
> 
> Doesn't that just mean this test is buggy, in that it was testing artificial behaviors produced by the test rather than testing what happens in real use?

I'd say the test _might_ be buggy and I was trying to be surgical in case it's not.

You're looking at the right code. The `null` checks (multiple since both the inner `Content` and the read stream are checked) are likely for tests and for wrapping `HttpRequestMessage`s but, as I said above, I don't understand the `HttpMessageContent` well enough to know the intent of the `CanSeek` check. I hesitate to change the product code given my lack of understanding.

The other part of `HttpMessageContent` I don't "get" is why the message headers are included in the `Length` calculation only when the `HttpResponseMessage.Content` (or `HttpRequestMessage.Content` though that's not relevant here) is `null` or the read stream checks succeed. `HttpContent` stops trying to calculate `Length` at that point.

Reply:
> I see for example that almost any interaction w/ the read stream in HttpContent will buffer it.

That's the base method.  It's overridden by every meaningful HttpContent-derived type.

> I hesitate to change the product code given my lack of understanding.

I'm not suggesting changing the product code.  I'm suggesting the test appears to be testing itself :), so I'm not sure what it's real purpose in life is.

Reply:
> > I see for example that almost any interaction w/ the read stream in HttpContent will buffer it.
> 
> That's the base method. It's overridden by every meaningful HttpContent-derived type.

That's not what I saw in `MultipartContent` or `ByteArrayContent` but I'm probably missing something and defer to you here.

> > I hesitate to change the product code given my lack of understanding.
> 
> I'm not suggesting changing the product code. I'm suggesting the test appears to be testing itself :), so I'm not sure what it's real purpose in life is.

Fair point. I _think_ the tests that break when running on `net6.0` (`SerializeResponse`, `SerializeResponseMultipleTimes`, and `SerializeResponseAsync`) are generally attempting to test how well `HttpMessageContent` handles an empty response message. It's certainly true the tests aren't constructing the real shape of such a response whether testing against `net6.0` or not. The question is how far off the tests are.

I'll check our TFS code to see if there's an analogous e2e test and debug that if so&hellip;

Reply:
@stephentoub after remembering how to build an ASP.NET / Web API web site and deploy it to IIS (:grin:), I created a simple console app using `HttpClient`. Every `HttpResponseMessage` returned from my site has `StreamContent` or `HttpConnectionResponseContent`. Every `content.ReadAsStreamAsync()` return value is a seekable, non-writeable `MemoryStream`. If it matters, my actions return empty responses &mdash; HTTP 202, 302, 400, or 500. For example,

![image](https://user-images.githubusercontent.com/6431421/216739393-900b8d40-7bdb-43b2-a759-2449f5c6ded1.png)

----

Things get interesting when using batching (one of two reasons I know of to use `HttpMessageContent`). Our [batching sample](https://github.com/aspnet/samples/tree/main/samples/aspnet/WebApi/BatchSample) clearly shows the client checking the individual `HttpResponseMessage.Content` values for `null`. And, when that's non-`null`, the code expects `await HttpResponseMessage.Content.ReadAsStringAsync()` to succeed.

Unfortunately, that last expectation is true only for `net481` and `netcoreapp2.1`. I didn't check every version, but the expectation fails for `net6.0` and `net7.0` because `Content` will never be (can never be) `null`. The issue here is that the customer code can now only use its expectations of an empty or non-empty response message to decide whether calling `ReadAsStringAsync()` could succeed. Newer .NET TFMs provide no external signal.

I can't think of a way to truly resolve this in this repo. My only thought would be to change the few places where we create `HttpResponeMessage`s to _always_ initialize its `Content`, perhaps using `new StringContent(string.Empty)` and clearing its `Headers`. Effectively, I'd throw away `EmptyContent` and make sure `ReadAsStringAsync()` works. But I expect this will cause problems (at least in our tests) because the serialization of those batched responses will change.

My general problem here is that aspnet/AspNetWebStack exposes explicitly-created `HttpResponseMessage`s to user code. This isn't what the dotnet/runtime change expected but it is the case.

Reply:
> one of two reasons I know of to use `HttpMessageContent`

/fyi the other is just a nice way to log an entire request or response message. If there are more use cases, I am not aware of them. And the large number of AspNetWebStack forks is preventing me from searching GitHub.

Reply:
> Every HttpResponseMessage returned from my site has StreamContent or HttpConnectionResponseContent. Every content.ReadAsStreamAsync() return value is a seekable, non-writeable MemoryStream

That's because you're calling `GetAsync(string)`, which uses the default HttpCompletionOption.ResponseContentRead; in other words, you're calling a method that explicitly buffers the entire response, which is why you're getting back the MemoryStream that contains that whole buffer.  Instead, call GetAsync with ResponseHeadersRead, and you'll see it not buffered, e.g.
```C#
using System;
using System.IO;
using System.Net.Http;
using System.Threading.Tasks;

internal class Program
{
    static async Task Main()
    {
        using (var client = new HttpClient())
        {
            HttpResponseMessage response = await client.GetAsync("http://httpbin.org", HttpCompletionOption.ResponseHeadersRead);
            Console.WriteLine(response.Content.GetType());
            using (Stream stream = await response.Content.ReadAsStreamAsync())
            {
                Console.WriteLine(stream.GetType());
                Console.WriteLine(stream.CanSeek);
            }
        }
    }
}
```
outputs:
```
System.Net.Http.StreamContent
System.Net.Http.StreamContent+ReadOnlyStream
False
```

> Our [batching sample](https://github.com/aspnet/samples/tree/main/samples/aspnet/WebApi/BatchSample) clearly shows the client checking the individual HttpResponseMessage.Content values for null.

Where does it show that?  I'm not sure how to be more explicit other than showing the code for HttpClientHandler in .NET Framework... here's the code that creates the response:
https://github.com/microsoft/referencesource/blob/dae14279dd0672adead5de00ac8f117dcf74c184/System/net/System/Net/Http/HttpClientHandler.cs#L1125-L1131
It's always set to a StreamContent and is literally never null.  What code is that BatchSample using where it expects HttpResponseMessage.Content to be null.

> the code expects await HttpResponseMessage.Content.ReadAsStringAsync() to succeed.

I continue to not understand the problem.  ReadAsStringAsync on an empty response message works just fine: it returns an empty string.

> I can't think of a way to truly resolve this in this repo.

Resolve what?  I don't understand what problem it is you're trying to solve.

Reply:
> My general problem here is that aspnet/AspNetWebStack exposes explicitly-created HttpResponseMessages to user code.

Where does it do that?

Reply:
> I continue to not understand the problem. ReadAsStringAsync on an empty response message works just fine: it returns an empty string.

I'm likely describing the problem incorrectly and need to test more to get the stack trace again.

More generally, the methods listed below behave differently in a fairly `public` way when run against `net6.0` or `net7.0`. They should behave consistently. Otherwise, how will users know the returned `HttpResponseMessage` must be special-cased❔

> Where does it do that?

1. In `HttpContentMessageExtensions.ReadAsHttpResponseMessageAsync(...)` methods
2. In `HttpRequestMessageExtensions.CreateResponse(...)` methods
3. In numerous other places outside the System.Net.Http.Formatting assemblies, all compiled to target only `net45`

Reply:
Took a bit of circling around because I'd tried something else and lost the problematic sequence&hellip;

I recreated the problem by changing my test client app to read from `response.Content.ReadAsMultipartAsync().Contents` members, displaying a few `HttpContent` members before using `ReadAsHttpResponseMessageAsync()`. That fails w/ **all** TFMs _except_ `net481`. So, I was mistaken in two ways: The symptoms occur when targeting `netcoreapp2.1` as well and the method wasn't `ReadAsStringAsync()`.

Weirdly the problem seems to arise only when my `DisplayContent(...)` method uses both `ReadAsStreamAsync()`, and `ReadAsStringAsync()`. Calling `ReadAsHttpResponseMessageAsync()` after that fails w/ the following stack trace. Calling one or the other seems to work fine.

``` text
Unhandled exception. System.IO.IOException: Unexpected end of HTTP message stream. HTTP message is not complete.
   at System.Net.Http.HttpContentMessageExtensions.ReadAsHttpResponseMessageAsyncCore(HttpContent content, Int32 bufferSize, Int32 maxHeaderSize, CancellationToken cancellationToken)
   at Program.<Main>$(String[] args) in C:\dd\Projects\NameValueCollection.115\ConsoleApp\Program.cs:line 33
   at Program.<Main>(String[] args)
```

----

Maybe the `IOException` above is for a corner case scenario. My broader concern remains: User code would previously see `HttpResponseMessage.Content == null` and could take that as an indication that (batched) response was empty. They no longer have that signal. Use of the methods mentioned above is not a test-only scenario.

Reply:
> More generally, the methods listed below behave differently in a fairly public way when run against net6.0 or net7.0. They should behave consistently. Otherwise, how will users know the returned HttpResponseMessage must be special-cased❔

We're going around in circles.  Can you please provide a standalone, complete runnable example that:
a) demonstrates the difference in behavior
b) is code that someone is expected to actually write and not just as part of a test
c) isn't just about showing that `new HttpResponseMessage().Content` returns null vs non-null but rather something higher level exhibiting wrong behavior because of it
?

As previously discussed, we took [a change](https://github.com/dotnet/runtime/pull/35910) in .NET 5 that made it so that `HttpResponseMessage.Content` doesn't return null.  It's a fact that the change can be observed, that's not in question.  However, consumer code is never supposed to be handed back an `HttpResponseMessage` that has a null Content, which is why the change was deemed ok to make, and not making that change, in particular in the presence of nullable reference types annotations, has far worse impact.  So the question here isn't whether code that does `new HttpResponseMessage().Content is null` can see a change in behavior... it absolutely can.  The question is where that actually matters, because it's not supposed to.

Thanks.

Reply:
I agree we're talking past each other @stephentoub, likely primarily because I have mixed a number of issues into the thread as I attempted to work around the original test failure in this PR. I'll try to unravel things a bit&hellip;

The bottom line is the runtime change in 5.0 was _not_ test-only (as was advertised) and will change the behaviour of public code in the Formatting package, leading to `Content!=null` when `null` means something on other TFMs or incorrect `ContentLength` calculations. My concerns about `Stream.Length` throwing (user violent tho' that may be, at least `CanSeek` is documented to allow it) and the odd (corner) case where `ReadAsHttpResponseMessageAsync()` throws are secondary to that visible change.

Worse, the AspNetWebStack code depends on the read `Stream` being either `null` or seekable, especially in the batched case. I just haven't demonstrated the `ContentLength` calculations failing outside tests yet.

### Details

> > Where does it do that?

> 1. In `HttpContentMessageExtensions.ReadAsHttpResponseMessageAsync(...)` methods
> 2. In `HttpRequestMessageExtensions.CreateResponse(...)` methods
> &hellip;

I don't see a way to change this code to behave appropriately (avoid returning a `HttpResponseMessage` w/ `Content!=null` and then provide a correct `HttpContent.Headers.ContentLength`) w/o causing breaking changes for users of our Formatting package. If there is something, I'm willing to change the src/ code but my preference would be to revert the runtime change or quirk its behaviour such that projects depending on this package would automatically have the quirk enabled (making `Content==null` consistent).

> consumer code is never supposed to be handed back an `HttpResponseMessage` that has a null Content

We have been exposing such `Content==null` cases to user code since long before .NET 1.0. My point isn't really about the history of course. It's about moving forward i.e. either reverting the 5.0 commit (in 6.0 and 7.0) because we have new information or finding a way to fix the Formatting code w/o breaking users who depend on either `Content==null` or always having a correct `ContentLength` value (in other words, those used to running against .NET Framework, `netcoreapp2.1`, or `netcoreapp3.1`).

As you can tell, I'm now of the opinion we need a product change somewhere.

> Can you please provide a standalone, complete runnable example&hellip;

https://github.com/dougbu/BatchTests is close but currently demonstrates the `IOException` rather than a failure simply because the provided read `Stream` isn't seekable. I may need to tweak the service somehow to remove the response's `Content-Length` header or the client to explicitly check a `HttpMessageContent.Headers.ContentLength` value.

I'm not quite sure why `Content!=null` for an empty batched response is _not_ a breaking change (on its own) from your perspective. Could you elaborate❔

### Side note

Not much cares about the `HttpCompletionOption.ResponseHeadersRead` case from what I can see. Why would user code request early termination of the `SendAsync(...)` then try to handle the response's Content❔

### Moving forward

If this long-ish message hasn't helped us, let's meet in Teams later this week.

Reply:
> We have been exposing such Content==null cases to user code since long before .NET 1.0

Which I believe is a bug.  Developers expect to be able to dereference Content without doing a null check.  Your Program.cs, for example, does just that.

> the AspNetWebStack code depends on the read Stream being either null or seekable

The only way a networked response via Stream becomes seekable is if it's buffered.  If AspNetWebStack for whatever reasons really needs it to be seekable and it's not, then it can buffer it itself.  From where I sit, the requirement that a response stream from a networking component be seekable sounds like a long-standing bug in the AspNetWebStack that should be fixed if you actually believe folks are going to run into it.

> I'm not quite sure why Content!=null for an empty batched response is not a breaking change (on its own) from your perspective. Could you elaborate❔

To quote my previous response:
"It's a fact that the change can be observed, that's not in question. However, consumer code is never supposed to be handed back an HttpResponseMessage that has a null Content, which is why the change was deemed ok to make, and not making that change, in particular in the presence of nullable reference types annotations, has far worse impact."

> Not much cares about the HttpCompletionOption.ResponseHeadersRead case from what I can see.

That is simply not the case.

> Why would user code request early termination of the SendAsync(...) then try to handle the response's Content❔

Because otherwise the call has to buffer _the entire response payload_ before handing back any of it to the consumer.  Code that cares about performance will use ResponseHeadersRead, or will use HttpClient.GetStreamAsync that implicitly behaves like ResponseHeadersRead.

> either reverting the 5.0 commit (in 6.0 and 7.0) because we have new information or finding a way to fix the Formatting code w/o breaking users who depend on either Content==null or always having a correct ContentLength value

I don't see us reverting that.  It would mean we believe Content can be null, in which case every use of .Content everywhere is going to start getting nullability warnings erroneously telling people they need to do a null check because we'd have to make the property `HttpContent? Content`.

> If this long-ish message hasn't helped us, let's meet in Teams later this week.

Sure.

cc: @karelz, @MihaZupan

Reply:
I debugged a fair amount today and found operations using `HttpContent` are not idempotent, especially in newer .NET Core releases. Had to create lots of either / or cases to avoid `Exception`s and weird "empty" content (despite a `ContentLenght==741`).

I assume most applications don't attempt more than one operation on most `HttpContent` instances. But it's unclear to me when different operations may cause problems for each other. Are the restrictions documented somewhere❔

In any case, my updated [`BatchTests`](https://github.com/dougbu/BatchTests) now avoids the `IOException`s mentioned in earlier comments as well as `Exception`s arising elsewhere.

Reply:
Resolving this conversation because I just removed the affected tests when compiling for `net6.0`. Will create another PR soon.